core/src/main/scala/kafka/coordinator/GroupCoordinator.scala

```diff
@@ -0,0 +1,635 @@
+    heartbeatPurgatory.checkAndComplete(memberKey)
+
+    // reschedule the next heartbeat expiration deadline
+    val newHeartbeatDeadline = member.latestHeartbeat + member.sessionTimeoutMs
+    val delayedHeartbeat = new DelayedHeartbeat(this, group, member, newHeartbeatDeadline, member.sessionTimeoutMs)
+    heartbeatPurgatory.tryCompleteElseWatch(delayedHeartbeat, Seq(memberKey))
+  }
+
+  private def removeHeartbeatForLeavingConsumer(group: GroupMetadata, member: MemberMetadata) {
+    member.isLeaving = true
+    val consumerKey = ConsumerKey(member.groupId, member.memberId)
+    heartbeatPurgatory.checkAndComplete(consumerKey)
+  }
+
+  private def addMemberAndRebalance(sessionTimeoutMs: Int,
+                                    protocols: List[(String, Array[Byte])],
+                                    group: GroupMetadata,
+                                    callback: JoinCallback) = {
+    val memberId = group.generateNextMemberId
+    val member = new MemberMetadata(memberId, group.groupId, sessionTimeoutMs, protocols)
+    member.awaitingJoinCallback = callback
+    group.add(member.memberId, member)
+    maybePrepareRebalance(group)
+    member
+  }
+
+  private def removeMember(group: GroupMetadata, member: MemberMetadata) {
+    group.remove(member.memberId)
+  }
+
+  private def updateMemberAndRebalance(group: GroupMetadata,
+                                       member: MemberMetadata,
+                                       protocols: List[(String, Array[Byte])],
+                                       callback: JoinCallback) {
+    member.supportedProtocols = protocols
+    member.awaitingJoinCallback = callback
+    maybePrepareRebalance(group)
+  }
+
+  private def maybePrepareRebalance(group: GroupMetadata) {
+    group synchronized {
+      if (group.canRebalance)
+        prepareRebalance(group)
+    }
+  }
+
+  private def prepareRebalance(group: GroupMetadata) {
+    // if any members are awaiting sync, cancel their request and have them rejoin
+    if (group.is(AwaitingSync)) {
+      for (member <- group.allMembers) {
+        if (member.awaitingSyncCallback != null) {
+          member.awaitingSyncCallback(Array.empty, Errors.REBALANCE_IN_PROGRESS.code)
+          member.awaitingSyncCallback = null
+        }
+      }
+    }
+
+    group.allMembers.foreach(_.assignment = null)
+    group.transitionTo(PreparingRebalance)
+    info("Preparing to restabilize group %s with old generation %s".format(group.groupId, group.generationId))
+
+    val rebalanceTimeout = group.rebalanceTimeout
+    val delayedRebalance = new DelayedRebalance(this, group, rebalanceTimeout)
+    val consumerGroupKey = ConsumerGroupKey(group.groupId)
+    rebalancePurgatory.tryCompleteElseWatch(delayedRebalance, Seq(consumerGroupKey))
+  }
+
+  private def onMemberFailure(group: GroupMetadata, member: MemberMetadata) {
+    trace("Member %s in group %s has failed".format(member.memberId, group.groupId))
+    removeMember(group, member)
```

Comment:
I think it's better the follow the same pattern as we do in handleXXX here, e.g. move the checking such as ```rebalancePurgatory.checkAndComplete()``` if the state is in ``RepareRebalance```, or ```maybePrepareRebalance ``` if state is in ```AwaitingSync``` or ```Stable```.